Pktline rewrite

aymanbagabas opened the pull request:
This pktline rewrite implements the protocol as a set of methods under the `pktline` package. This makes it easy to incorporate single pktline read/write operations in a more complex scenario and tasks.

TODO:
- [x] Remove pktline.Scanner and pktline.Encoder

Benchmarks against the current pktline implementation. `Scanner/Encoder` is the current implementation, compared to this approach using for loops.

Note: `BenchmarkEncoder/empty-16` is different from `BenchmarkWritePacket/empty-16`. `pktline.Encoder` writes a flush packet on empty input, while `pktline.WritePacket` writes an empty packet `0004` on empty input. Thus the allocation and benchmark differ.

```
› go test -benchmem -bench .  -run=^$
goos: darwin
goarch: amd64
pkg: github.com/go-git/go-git/v5/plumbing/format/pktline
cpu: Intel(R) Core(TM) i9-9980HK CPU @ 2.40GHz
BenchmarkScanner/empty-16       39084597                27.66 ns/op            4 B/op          1 allocs/op
BenchmarkScanner/one_message-16                 40376830                28.13 ns/op            4 B/op          1 allocs/op
BenchmarkScanner/two_messages-16                40645405                30.10 ns/op            4 B/op          1 allocs/op
BenchmarkScanner/sections-16                    44230746                28.16 ns/op            4 B/op          1 allocs/op
BenchmarkReadPacket/empty-16                    20508476                52.33 ns/op            4 B/op          1 allocs/op
BenchmarkReadPacket/one_message-16              21085110                51.20 ns/op            4 B/op          1 allocs/op
BenchmarkReadPacket/two_messages-16             20531574                52.34 ns/op            4 B/op          1 allocs/op
BenchmarkReadPacket/sections-16                 24020688                51.73 ns/op            4 B/op          1 allocs/op
BenchmarkEncoder/empty-16                       65896110                18.53 ns/op            8 B/op          0 allocs/op
BenchmarkEncoder/one_message-16                 16062358                70.66 ns/op           61 B/op          2 allocs/op
BenchmarkEncoder/two_messages-16                16700629                84.41 ns/op           92 B/op          2 allocs/op
BenchmarkEncoder/sections-16                    12602998               118.6 ns/op           198 B/op          2 allocs/op
BenchmarkWritePacket/empty-16                   29277804                40.99 ns/op           33 B/op          1 allocs/op
BenchmarkWritePacket/one_message-16             17538825                66.95 ns/op           58 B/op          2 allocs/op
BenchmarkWritePacket/two_messages-16            16326508                72.26 ns/op           93 B/op          2 allocs/op
BenchmarkWritePacket/sections-16                12842865               215.3 ns/op           362 B/op          2 allocs/op
PASS
ok      github.com/go-git/go-git/v5/plumbing/format/pktline     23.363s


```

aymanbagabas committed:
plumbing: pktline, dry parsing length and rename empty pkt

aymanbagabas committed:
plumbing: pktline, implement pktline reader

aymanbagabas committed:
plumbing: pktline, implement pktline writer

aymanbagabas committed:
plumbing: pktline tests, add benchmarks

aymanbagabas committed:
plumbing: pktline, deprecate the old impl and use pktline read/writer

aymanbagabas commented:
Would love your feedback on this 🙂 

cc/ @pjbgf @blmayer

pjbgf commented on plumbing/format/pktline/common.go:
```diff
@@ -0,0 +1,36 @@
+
+// Status represents the status of a pktline. Any value greater than 4 is
+// considered a data pkt.
+type Status = int
```
An easy link to upstream documentation could be useful here.

aymanbagabas replied:
We probably don't need this, thinking about removing this type. It was modeled after Git `packet_read_status` https://github.com/git/git/blob/master/pkt-line.h#L113

pjbgf commented on plumbing/format/pktline/error.go:
```diff
@@ -30,18 +30,21 @@ func (e *ErrorLine) Error() string {
 func (e *ErrorLine) Encode(w io.Writer) error {
-	p := NewEncoder(w)
-	return p.Encodef("%s%s\n", string(errPrefix), e.Text)
+	p := NewWriter(w)
```
Maybe we could force `w` to be of the type we expect so we don't to force each call to `Encode` to cause at least one additional allocation.

pjbgf commented on plumbing/format/pktline/error.go:
```diff
@@ -30,18 +30,21 @@ func (e *ErrorLine) Error() string {
-	s := NewScanner(r)
-	if !s.Scan() {
-		return s.Err()
+	s := NewReader(r)
```
Same here on the allocation.

pjbgf commented on plumbing/format/pktline/reader.go:
```diff
@@ -0,0 +1,192 @@
+// NewReader returns a new pktline reader that reads from r and supports
+// peeking.
+func NewReader(r io.Reader) *Reader {
+	if rdr, ok := r.(*Reader); ok {
```
Maybe two different `New` funcs, which are type specific so we can avoid the type cast check.

blmayer replied:
Should NewReader return the io.Reader interface? Since the reader struct implements the interface. But we'll have a different architecture in this case. Makes sense?

aymanbagabas replied:
We no longer need the reader/writer nor scanner/encoder, i will remove those before merging

pjbgf commented on plumbing/format/pktline/reader.go:
```diff
@@ -0,0 +1,192 @@
+	buf := r.buf[lenSize : lenSize+dataLen]
+	if bytes.HasPrefix(buf, errPrefix) {
+		err = &ErrorLine{
+			Text: strings.TrimSpace(string(buf[4:])),
```
```suggestion
			Text: string(bytes.TrimSpace(buf[4:])),
```

aymanbagabas committed:
plumbing: pktline, use static methods approach

aymanbagabas commented:
Now, we don't need a reader/writer scanner/encoder to read/write pktline. Use existing io.Reader/io.Writer along with `pktline.ReadPacket/pktline.WritePacket`

blmayer commented on plumbing/format/pktline/writer.go:
```diff
@@ -0,0 +1,55 @@
+}
+
+// WriteDelim writes a delimiter packet.
+func (w *Writer) WriteDelim() (err error) {
```
I think you're missing the WriteResponseEnd method to write the 0002 packet.

blmayer commented on plumbing/format/pktline/pktline.go:
```diff
@@ -0,0 +1,198 @@
+}
+
+// WriteDelim writes a delimiter packet.
+func WriteDelim(w io.Writer) (err error) {
```
What about writing the response end packet?

blmayer reviewed (commented):
Hi, I like the way things are going here, letting people use the io interfaces instead of the encoder looks simpler. Great work here, just left some comments about the responseEnd packet.

aymanbagabas committed:
plumbing: pktline, add WriteResponseEnd and method docs

blmayer commented:
I added a comment before approving. Great job.

pjbgf commented on plumbing/format/pktline/pktline.go:
```diff
@@ -0,0 +1,225 @@
+
+// WritePacket writes a pktline packet.
+func WritePacket(w io.Writer, p []byte) (n int, err error) {
+	defer func() {
```
Let's add a `nil` check for `w` - as this would be publicly exposed.

pjbgf commented on plumbing/protocol/packp/uppackresp.go:
```diff
@@ -17,6 +16,7 @@ var ErrUploadPackResponseNotDecoded = errors.New("upload-pack-response should be
 // UploadPackResponse contains all the information responded by the upload-pack
 // service, the response implements io.ReadCloser that allows to read the
 // packfile directly from it.
+// TODO: to be removed
```
Before this PR merges or at a given milestone?

aymanbagabas replied:
At v6, i'm planning on deconstructing this and moving it to `transport` since `UploadPackResponse` is not a _pack protocol` data structure. Probably rename it to `FetchResponse`

pjbgf replied:
```suggestion
// TODO: v6, to be removed
```

pjbgf commented on plumbing/protocol/packp/srvresp.go:
```diff
@@ -20,20 +21,25 @@ type ServerResponse struct {
-func (r *ServerResponse) Decode(reader *bufio.Reader, isMultiACK bool) error {
-	s := pktline.NewScanner(reader)
+func (r *ServerResponse) Decode(reader io.Reader, isMultiACK bool) error {
+	s := bufio.NewReader(reader)
```
We could save the additional allocation by receiving this as `reader` instead - so keeping the previous API and ditching this line instead.

aymanbagabas replied:
This is fine for now. I'm planning to refactor this bit and remove the need for a `bufio.Reader` in the transport refactor. Right now, it is needed to _peek_ into the next packets to determine if the client reached a packfile. This should be done in the `transport` pack negotiation layer and not here. See https://git-scm.com/docs/pack-protocol#_packfile_negotiation

pjbgf commented on plumbing/format/pktline/scanner_test.go:
```diff
@@ -44,7 +43,7 @@ func (s *SuiteScanner) TestDecodeOversizePktLines(c *C) {
 		sc := pktline.NewScanner(r)
 		_ = sc.Scan()
-		c.Assert(sc.Err(), IsNil)
+		c.Assert(sc.Err(), NotNil)
```
We may want to add a comment on Scan() to highlight why previously we would not get an error and now we do.

pjbgf commented on plumbing/format/pktline/reader.go:
```diff
@@ -0,0 +1,93 @@
+	}
+
+	readLen := n - len(r.buf)
+	readBuf := make([]byte, readLen)
```
Maybe the memory churn could be alliviated using `sync.GetBytSlice()` instead of creating a new byte array.

aymanbagabas replied:
We need to create new byte arrays to make this concurrent safe, especially if the caller will modify the returned packet slice. Using `sync.GetByteSlice/PutByteSlice` can solve the concurrency problem but it introduces more overhead i.e. greater than `55 ns/op` at least on my machine.

The `Scanner` implementation assumes that there will be one caller at a time, and it's the caller's responsibility to _copy_ the packet slice from `s.Bytes()`. While this is efficient, it's not stateless and depends on a `Scanner` to perform the reading.

pjbgf commented on plumbing/format/pktline/reader.go:
```diff
@@ -0,0 +1,93 @@
+		return nil, err
+	}
+
+	r.buf = append(r.buf, readBuf[:readN]...)
```
A call to [slices.Grow](https://pkg.go.dev/golang.org/x/exp/slices#Grow) would ensure a single op on the backing array.

aymanbagabas replied:
I've removed the `Reader` type, using a stateless implementation is IMO more composable and abstract

aymanbagabas committed:
plumbing: pktline, remove scanner/encoder

aymanbagabas committed:
plumbing: pktline, update comment

aymanbagabas commented:
@pjbgf could you give this another look since much has changed?

aymanbagabas committed:
plumbing: pktline, add benchmark tests

aymanbagabas committed:
plumbing: pktline, error on nil writer

pjbgf commented on plumbing/protocol/packp/report_status.go:
```diff
@@ -86,19 +90,23 @@ func (s *ReportStatus) Decode(r io.Reader) error {
 	}
 
-	return scan.Err()
+	if err != nil && err != io.EOF {
```
```suggestion
	if err != nil && !errors.Is(err, io.EOF) {
```

pjbgf commented on plumbing/protocol/packp/report_status.go:
```diff
@@ -86,19 +90,23 @@ func (s *ReportStatus) Decode(r io.Reader) error {
-		return nil
+func (s *ReportStatus) scanFirstLine(r io.Reader) ([]byte, error) {
+	_, p, err := pktline.ReadPacket(r)
+	if err == io.EOF {
```
```suggestion
	if errors.Is(err, io.EOF) {
```

pjbgf commented on plumbing/protocol/packp/gitproto.go:
```diff
@@ -74,16 +73,15 @@ func (g *GitProtoRequest) Encode(w io.Writer) error {
-			return ErrInvalidGitProtoRequest
-		}
+	_, p, err := pktline.ReadPacket(r)
+	if err == io.EOF {
```
```suggestion
	if errors.Is(err, io.EOF) {
```

pjbgf commented on plumbing/format/pktline/pktline_bench_test.go:
```diff
@@ -0,0 +1,91 @@
+				}
+			}
+		})
+	}
```
I believe we want to reset the reader before each call:

```suggestion
	for _, tc := range cases {
		r := strings.NewReader("")
		b.Run(tc.name, func(b *testing.B) {
			for i := 0; i < b.N; i++ {
				r.Reset(tc.input)
				_, _, err := pktline.ReadPacket(r)
				if err != nil {
					break
				}
			}
		})
	}
```

This results on more sensible benchmarks - note the gap between an empty input and the rest:

```
goos: linux
goarch: amd64
pkg: github.com/go-git/go-git/v5/plumbing/format/pktline
cpu: 11th Gen Intel(R) Core(TM) i7-11800H @ 2.30GHz
BenchmarkReadPacket/empty-16         	1000000000	         0.0000005 ns/op	       0 B/op	       0 allocs/op
BenchmarkReadPacket/one_message-16   	12724945	        82.79 ns/op	      40 B/op	       3 allocs/op
BenchmarkReadPacket/two_messages-16  	14444020	        83.35 ns/op	      40 B/op	       3 allocs/op
BenchmarkReadPacket/sections-16      	13942911	        86.68 ns/op	      40 B/op	       3 allocs/op
```

pjbgf commented on plumbing/format/pktline/pktline.go:
```diff
@@ -0,0 +1,229 @@
+func PeekPacket(r ioutil.ReadPeeker) (l int, p []byte, err error) {
+	defer func() {
+		if err == nil {
+			trace.Packet.Printf("packet: < %04x %s", l, p)
```
This is adding an additional allocation even when trace is disabled - which is quite costly.
Just for testing purposes, I made this changes which resulted in no additional alloc:

```suggestion
			trace.Packet.Printfunc(func() string { return fmt.Sprintf("packet: < %04x %s", l, p) })
```
And adding on `trace`:
```go
func (t Target) Printfunc(data func() string) {
	if int32(t)&current.Load() != 0 {
		logger.Output(2, data()) // nolint: errcheck
	}
}
```

We may need to think a bit more on ways around this to keep a nice API while dodging this additional allocation.

pjbgf commented on plumbing/format/pktline/pktline.go:
```diff
@@ -0,0 +1,229 @@
+// Use packet length to determine the type of packet i.e. 0 is a flush packet,
+// 1 is a delim packet, 2 is a response-end packet, and a length greater or
+// equal to 4 is a data packet.
+func ReadPacket(r io.Reader) (l int, p []byte, err error) {
```
As `ReadPacket` is static we have "limited" options on avoiding unecessary allocations. Currently it has two sources for allocation: `pktlen` and `data`. 

An option here would be able to provide a buffer and rely on a `io.LimitedReader` instead of `io.Reader`. Also adding the trace "fix":
```go
func ReadPacketBuf(lr *io.LimitedReader, buf *bytes.Buffer) (l int, p []byte, err error) {
	defer func() {
		if err == nil {
			trace.Packet.Printfunc(func() string { return fmt.Sprintf("packet: < %04x %s", l, p) })
		}
	}()

	lr.N = 4
	buf.Reset()
	buf.Grow(4)

	n, err := io.Copy(buf, lr)
	if err != nil {
		if errors.Is(err, io.ErrUnexpectedEOF) {
			return Err, nil, fmt.Errorf("%w: %d", ErrInvalidPktLen, n)
		}

		return Err, nil, err
	}

	if n != lenSize {
		return Err, nil, fmt.Errorf("%w: %d", ErrInvalidPktLen, n)
	}

	length, err := ParseLength(buf.Bytes())
	if err != nil {
		return Err, nil, err
	}

	switch length {
	case Flush, Delim, ResponseEnd:
		return length, nil, nil
	case lenSize: // empty line
		return length, []byte{}, nil
	}

	buf.Reset()
	dataLen := length - lenSize
	buf.Grow(dataLen)

	lr.N = int64(dataLen)
	dn, err := io.Copy(buf, lr)
	if err != nil {
		return Err, nil, err
	}

	if dn != int64(dataLen) {
		return Err, nil, fmt.Errorf("%w: %d", ErrInvalidPktLen, dn)
	}

	data := buf.Bytes()
	if bytes.HasPrefix(data, errPrefix) {
		err = &ErrorLine{
			Text: string(bytes.TrimSpace(data[4:])),
		}
	}

	return length, data, err
}
```

That results in no allocations:
```
BenchmarkReadPacketBuf/empty-16         	1000000000	         0.0000042 ns/op	       0 B/op	       0 allocs/op
BenchmarkReadPacketBuf/one_message-16   	16635014	        74.15 ns/op	       0 B/op	       0 allocs/op
BenchmarkReadPacketBuf/two_messages-16  	16032168	        72.44 ns/op	       0 B/op	       0 allocs/op
BenchmarkReadPacketBuf/sections-16      	16782122	        75.37 ns/op	       0 B/op	       0 allocs/op
```

aymanbagabas replied:
Thanks for the review @pjbgf. I think you're onto something here, we could improve this and make it more generic by using `ReadPacketBuf(io.Reader, []byte)` instead which makes it closer to the upstream `int packet_read` method signature.

Take a look at this implementation:
```go

func ReadPacketBuf(r io.Reader, p []byte) (l int, err error) {
	defer func() {
		if err == nil {
			trace.Packet.Printfunc(func() string { return fmt.Sprintf("packet: < %04x %s", l, p) })
		}
	}()

	if len(p) < lenSize {
		return Err, fmt.Errorf("%w: %d", ErrInvalidPktLen, len(p))
	}

	n, err := r.Read(p[:lenSize])
	if err != nil {
		return Err, err
	}

	if n != lenSize {
		return Err, fmt.Errorf("%w: %d", ErrInvalidPktLen, n)
	}

	length, err := ParseLength(p)
	if err != nil {
		return Err, err
	}

	switch length {
	case Flush, Delim, ResponseEnd:
		return length, nil
	case lenSize: // empty line
		return length, nil
	}

	if len(p) < length {
		return Err, fmt.Errorf("%w: %d", ErrInvalidPktLen, len(p))
	}

	dataLen := length - lenSize
	dn, err := r.Read(p[:dataLen])
	if err != nil {
		return Err, err
	}

	if dn != dataLen {
		return Err, fmt.Errorf("%w: %d", ErrInvalidPktLen, dn)
	}

	if bytes.HasPrefix(p, errPrefix) {
		err = &ErrorLine{
			Text: string(bytes.TrimSpace(p[4:])),
		}
	}

	return length, err
}
```

This results in zero allocs and a _slight_ improvement over the current `Scanner` implementation:
```
› go test -bench=. -benchmem
goos: darwin
goarch: amd64
pkg: github.com/go-git/go-git/v5/plumbing/format/pktline
cpu: Intel(R) Core(TM) i9-9980HK CPU @ 2.40GHz
BenchmarkScanner/empty-16       90463454                13.10 ns/op            0 B/op          0 allocs/op
BenchmarkScanner/one_message-16                 38528110                28.36 ns/op            0 B/op          0 allocs/op
BenchmarkScanner/two_messages-16                41303529                28.42 ns/op            0 B/op          0 allocs/op
BenchmarkScanner/sections-16                    41827941                28.50 ns/op            0 B/op          0 allocs/op
BenchmarkReadPacket/empty-16                    100000000               10.08 ns/op            0 B/op          0 allocs/op
BenchmarkReadPacket/one_message-16              45403502                25.11 ns/op            0 B/op          0 allocs/op
BenchmarkReadPacket/two_messages-16             43133918                25.16 ns/op            0 B/op          0 allocs/op
BenchmarkReadPacket/sections-16                 42402376                25.03 ns/op            0 B/op          0 allocs/op
BenchmarkEncoder/empty-16                       33795769                36.49 ns/op           19 B/op          1 allocs/op
BenchmarkEncoder/one_message-16                 27409754                43.34 ns/op           43 B/op          1 allocs/op
BenchmarkEncoder/two_messages-16                33945912                49.15 ns/op           67 B/op          1 allocs/op
BenchmarkEncoder/sections-16                    16685656               132.9 ns/op           261 B/op          1 allocs/op
BenchmarkWritePacket/empty-16                   94312168                12.10 ns/op           11 B/op          0 allocs/op
BenchmarkWritePacket/one_message-16             29564037                34.53 ns/op           40 B/op          1 allocs/op
BenchmarkWritePacket/two_messages-16            35636349                39.47 ns/op           64 B/op          1 allocs/op
BenchmarkWritePacket/sections-16                17210281                67.78 ns/op          253 B/op          1 allocs/op
PASS
ok      github.com/go-git/go-git/v5/plumbing/format/pktline     25.918s
```

You can find the changes I made to generate the benchmarks in this commit https://github.com/aymanbagabas/go-git/commit/33dc6724c4edbbd88a6c46a3e2b5ab730061d422

aymanbagabas committed:
plumbing: pktline, update plumbing/protocol/packp/uppackresp.go

Co-authored-by: Paulo Gomes <paulo.gomes.uk@gmail.com>

aymanbagabas committed:
plumbing: pktline, update plumbing/protocol/packp/report_status.go

Co-authored-by: Paulo Gomes <paulo.gomes.uk@gmail.com>

aymanbagabas committed:
plumbing: pktline, update plumbing/protocol/packp/report_status.go

Co-authored-by: Paulo Gomes <paulo.gomes.uk@gmail.com>

aymanbagabas committed:
plumbing: pktline, update plumbing/protocol/packp/gitproto.go

Co-authored-by: Paulo Gomes <paulo.gomes.uk@gmail.com>

aymanbagabas committed:
plumbing: pktline, update plumbing/format/pktline/pktline_bench_test.go

Co-authored-by: Paulo Gomes <paulo.gomes.uk@gmail.com>

aymanbagabas committed:
plumbing: packp, update imports

aymanbagabas committed:
plumbing: pktline, add GetPacketBuffer and PutPacketBuffer

aymanbagabas committed:
plumbing: pktline, use shared buffer with ReadPacketLine

aymanbagabas committed:
plumbing: pktline, add scanner based type

aymanbagabas committed:
plumbing: pktline, update benchmarks

aymanbagabas commented:
@pjbgf I've updated the API to separate the `ReadPacket` logic to make it allocation independent and added back a `Scanner` that pre-allocates a 65520 (65kb) slice for packets. Here're some benchmarks:

```
› go test -bench=. -benchmem
goos: darwin
goarch: amd64
pkg: github.com/go-git/go-git/v5/plumbing/format/pktline
cpu: Intel(R) Core(TM) i9-9980HK CPU @ 2.40GHz
BenchmarkScannerOld/empty-16    71351804                15.86 ns/op            0 B/op          0 allocs/op
BenchmarkScannerOld/one_message-16              27927124                39.93 ns/op            0 B/op          0 allocs/op
BenchmarkScannerOld/two_messages-16             16620072                67.24 ns/op            0 B/op          0 allocs/op
BenchmarkScannerOld/sections-16                  4122507               289.9 ns/op             0 B/op          0 allocs/op
BenchmarkScannerOld/max_packet_size-16            873728              1353 ns/op               0 B/op          0 allocs/op
BenchmarkScanner/empty-16                       100000000               11.31 ns/op            0 B/op          0 allocs/op
BenchmarkScanner/one_message-16                 28336495                38.52 ns/op            0 B/op          0 allocs/op
BenchmarkScanner/two_messages-16                18215517                66.81 ns/op            0 B/op          0 allocs/op
BenchmarkScanner/sections-16                     4142722               304.0 ns/op             0 B/op          0 allocs/op
BenchmarkScanner/max_packet_size-16               821563              1304 ns/op               0 B/op          0 allocs/op
BenchmarkReadPacket/empty-16                    88745823                12.10 ns/op            0 B/op          0 allocs/op
BenchmarkReadPacket/one_message-16              29064722                38.36 ns/op            0 B/op          0 allocs/op
BenchmarkReadPacket/two_messages-16             17287957                65.38 ns/op            0 B/op          0 allocs/op
BenchmarkReadPacket/sections-16                  4315484               280.5 ns/op             0 B/op          0 allocs/op
BenchmarkReadPacket/max_packet_size-16            848632              1289 ns/op               0 B/op          0 allocs/op
BenchmarkReadPacketLine/empty-16                40930602                27.41 ns/op            0 B/op          0 allocs/op
BenchmarkReadPacketLine/one_message-16          16346572                71.20 ns/op            0 B/op          0 allocs/op
BenchmarkReadPacketLine/two_messages-16         10143254               120.3 ns/op             0 B/op          0 allocs/op
BenchmarkReadPacketLine/sections-16              2579047               441.5 ns/op             0 B/op          0 allocs/op
BenchmarkReadPacketLine/max_packet_size-16        838102              1316 ns/op               0 B/op          0 allocs/op
BenchmarkWritePacket/empty-16                   23840467                47.86 ns/op           35 B/op          1 allocs/op
BenchmarkWritePacket/one_message-16             15424958                74.63 ns/op           62 B/op          2 allocs/op
BenchmarkWritePacket/two_messages-16            15968304                87.92 ns/op           95 B/op          2 allocs/op
BenchmarkWritePacket/sections-16                11517386               177.0 ns/op           214 B/op          2 allocs/op
PASS
ok      github.com/go-git/go-git/v5/plumbing/format/pktline     31.478s
```

You can see there is a _slight_ improvement over the original scanner implementation which is a good thing. You can find the code used to generate these benchmarks here 457604067359

Note that these benchmarks were collected with tracing commented out, we need to find a way to trace packets without allocating memory as you mentioned.

EDIT: updated the benchmarks and fixed tests

aymanbagabas committed:
plumbing: pktline, fix errors and improve perf and memory

pjbgf commented on plumbing/protocol/packp/advrefs_decode.go:
```diff
@@ -19,12 +19,12 @@ func (a *AdvRefs) Decode(r io.Reader) error {
-	hash  plumbing.Hash    // last hash read
-	err   error            // sticky error, use the parser.error() method to fill this out
-	data  *AdvRefs         // parsed data is stored here
+	s     io.Reader     // a pkt-line scanner from the input stream
```
```suggestion
	s     io.Reader     // a pkt-line reader from the input stream
```

pjbgf commented:
Here's some E2E benchmarks with impact on PlainClone between `go-git/master` vs this changes:

```
goos: linux
goarch: amd64
pkg: github.com/go-git/go-git/v5
cpu: 11th Gen Intel(R) Core(TM) i7-11800H @ 2.30GHz
              │ /tmp/before │          /tmp/after           │
              │   sec/op    │    sec/op     vs base         │
PlainClone-16   622.2m ± 9%   585.5m ± 13%  ~ (p=0.382 n=8)

              │ /tmp/before  │          /tmp/after           │
              │     B/op     │     B/op      vs base         │
PlainClone-16   1.699Mi ± 6%   1.623Mi ± 5%  ~ (p=0.083 n=8)

              │ /tmp/before │          /tmp/after          │
              │  allocs/op  │  allocs/op   vs base         │
PlainClone-16   26.24k ± 0%   26.15k ± 1%  ~ (p=0.442 n=8)
```

Allocations and memory churn look good on very large repositories (linux kernel) with slow clones:

```
goos: linux
goarch: amd64
pkg: github.com/go-git/go-git/v5
cpu: 11th Gen Intel(R) Core(TM) i7-11800H @ 2.30GHz
              │ /tmp/before │          /tmp/after          │
              │   sec/op    │   sec/op     vs base         │
PlainClone-16   40.73 ± ∞ ¹   42.49 ± ∞ ¹  ~ (p=0.686 n=4)
¹ need >= 6 samples for confidence interval at level 0.95

              │  /tmp/before  │           /tmp/after           │
              │     B/op      │     B/op       vs base         │
PlainClone-16   506.9Mi ± ∞ ¹   468.7Mi ± ∞ ¹  ~ (p=0.343 n=4)
¹ need >= 6 samples for confidence interval at level 0.95

              │ /tmp/before  │             /tmp/after             │
              │  allocs/op   │  allocs/op    vs base              │
PlainClone-16   3.646M ± ∞ ¹   3.613M ± ∞ ¹  -0.90% (p=0.029 n=4)
¹ need >= 6 samples for confidence interval at level 0.95
```

Based on the above, and the previous discussions, LGTM.

aymanbagabas committed:
plumbing: pktline, shorten method names

Pktline is already implied in the package name, no need to have "Packet"
in the method names.

aymanbagabas committed:
plumbing: pktline, update plumbing/protocol/packp/advrefs_decode.go

Co-authored-by: Paulo Gomes <paulo.gomes.uk@gmail.com>
